### PR TITLE
Add basic support for building C via swiftpm

### PR DESCRIPTION
**Pull request by aciidb0mb3r**
Basic implementation of proposal https://github.com/apple/swift-evolution/blob/master/proposals/0038-swiftpm-c-language-targets.md

* Swiftpm can be used to create and build C libraries.
* Same rules apply for layouts except that `include/module.modulemap` should be present for each C lang module
* If a dir contains `.swift` and `.c`, it will error out.
* Right now only creates shared libraries and not executable.
* Assumes `module.modulemap` will be available inside a `include` directory, doesn't generates it automatically.
* Adds Test cases for possible layouts and mixing of swift and c code.

Remaining features will be added incrementally.

### CODE REVIEW

File: Fixtures/CLangModules/CLibraryFlat/Foo.c [hunk 1]
```diff
@@ -0,0 +1,6 @@
+int foo() {
```
Comment (gribozavr) [inline Fixtures/CLangModules/CLibraryFlat/Foo.c]: 'Clang' is spelled with a lowercase 'l'.

Comment (aciidb0mb3r) [inline Fixtures/CLangModules/CLibraryFlat/Foo.c]: Actually I wanted to name it CModules but CModules refers to system libraries having a modulemaps as a wrapper. so I changed it to C Lang which spelled CLang lol, but anyway Clang would make perfect sense here too. Changing.

**Comment by aciidb0mb3r** [comment]
@mxcl @ddunbar would love inputs on this

**Comment by mxcl** [comment]
Looks like a great start and really: great job getting this going.

I'll review it more thoroughly in the morning.

**Comment by ddunbar** [comment]
OMG awesome! Will review ASAP

**Comment by ddunbar** [comment]
Overall seems like a great start to me. There are a couple things I'd like to see fixed before merge (mostly having to do with erroring out on unsupported things, or not erroring out on things which do work but aren't yet complete (like not require the module.modulemap to be written, even though we don't synthesize it yet)), but once @mxcl is happy with it I'd be glad to see it go in and then build in the other things like module map synthesis or incremental build support.

**Comment by aciidb0mb3r** [comment]
@ddunbar I think I was able to resolve almost all of the things pointed by you. You can re-review the changes here: https://github.com/apple/swift-package-manager/pull/183/files

File: Sources/Build/describe().swift [hunk 2]
```diff
@@ -103,6 +103,65 @@ public func describe(prefix: String, _ conf: Configuration, _ modules: [Module],
...
+        }
+        
+        let inputs = module.dependencies.map{ $0.targetName } + module.sources.paths
+        let productPath = Path.join(prefix, "lib\(module.c99name).so")
+        let wd = Path.join(prefix, "\(module.c99name).build")
+        mkdirs.insert(wd)
+        
+        var args: [String] = []
+    #if os(Linux)
+         args += ["-fPIC"]
```
Comment (aciidb0mb3r) [inline Sources/Build/describe().swift]: Tests were failing on linux and clang suggested to recompile using this. Not sure if this is correct? @mxcl @ddunbar

**Comment by aciidb0mb3r** [comment]
Also I am waiting for a new toolchain to arrive before I resolve conflicts.

File: Sources/Build/describe().swift (hunk 2 above)
Comment (ddunbar) [inline Sources/Build/describe().swift]: That is probably correct behavior, I would want to test it with Swift interop to know for sure.

Comment (aciidb0mb3r) [inline Sources/Build/describe().swift]: I've tests in this PR calling c methods from swift.

**Comment by aciidb0mb3r** [comment]
I succeeded in building toolchain from latest masters and then use that for developing this. Now rebased correctly. 😬

**Comment by aciidb0mb3r** [comment]
@mxcl CI on this? though probably review first.

**Comment by ddunbar** [comment]
Let's try it.

@swift-ci please test

**Comment by mxcl** [comment]
A key point in our evolution, many thanks everyone 👊🏻

**Comment by ddunbar** [comment]
Woot!

**Comment by aciidb0mb3r** [comment]
😱😱 didn't expect a merge so soon. Thanks! It was fun doing this. Will try adding other remaining things from proposal😀

File: Sources/Transmute/Package+modules.swift [hunk 3]
```diff
@@ -74,19 +74,36 @@ extension Package {
...
+    func modulify(path: String, name: String) throws -> Module {
+        let walked = walk(path, recursing: shouldConsiderDirectory).map{ $0 }
+        
+        let cSources = walked.filter{ isValidSource($0, validExtensions: Sources.validCExtensions) }
+        let swiftSources = walked.filter{ isValidSource($0, validExtensions: Sources.validSwiftExtensions) }
+        
+        if !cSources.isEmpty {
+            guard swiftSources.isEmpty else { throw Module.Error.MixedSources(path) }
+            //FIXME: Support executables for C languages
+            guard !cSources.contains({ $0.hasSuffix("main.c") }) else { throw Module.Error.CExecutableNotSupportedYet(path) }
```
Comment (mxcl) [inline Sources/Transmute/Package+modules.swift]: Should check the basename.

Comment (aciidb0mb3r) [inline Sources/Transmute/Package+modules.swift]: raised #199 for this

**Pull request merged**